Web Inspector: The new ServiceWorkerDebuggableProxy needs to adopt auto inspection

the-chenergy opened the pull request:
#### 50e5a0f85821c2e0ff2373edab9b7f8ce01d997e
<pre>
Web Inspector: The new ServiceWorkerDebuggableProxy needs to adopt auto inspection
<a href="https://rdar.apple.com/145795990">rdar://145795990</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=288776">https://bugs.webkit.org/show_bug.cgi?id=288776</a>

Reviewed by Patrick Angle and Per Arne Vollan.

The work in [1] introduced the ServiceWorkerDebuggableProxy that
provides debugging support to service worker from the UI process.
This commit migrates the work done in [2] to support automatically
inspecting this new debuggable type instead.

- [1] <a href="https://github.com/WebKit/WebKit/pull/40437">https://github.com/WebKit/WebKit/pull/40437</a> (Add ServiceWorkerDebuggableProxy)
- [2] <a href="https://github.com/WebKit/WebKit/pull/20316">https://github.com/WebKit/WebKit/pull/20316</a> (Enable auto-inspecting ServiceWorkerDebuggable)

To summarize how automatic inspection generally works:
1. The web content marks its debuggable as inspectable.
2. The debuggable blocks its thread before starting code execution.
3. The auto-launched frontend, after initializing, notifies the backend.
4. The backend unblocks the debuggable to resume its code execution.

The ServiceWorkerDebuggableProxy lives in the UI process but controls
the service worker&apos;s thread. Notably, during step 3, the UI process is
also the one handling responses from the remote inspector, meaning the
UI process can&apos;t be blocked while waiting for the frontend. Instead, it
tells the underlying service worker to remain paused. The web process
then carries on with the normal procedure of waiting for the frontend&apos;s
signal.

* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
(WebKit::WebProcessProxy::createServiceWorkerDebuggable):
* Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.cpp:
(WebKit::ServiceWorkerDebuggableProxy::pauseWaitingForAutomaticInspection):
   - In step 2, when we mark the debuggable as inspectable, it should
     pause and prevent code execution. This is now done as an async
     response to tell the service worker thread to remain paused from
     the WaitForInspector start mode.

* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::installServiceWorker):
   - To cooperate and finish step 2, handle the request to either pause
     or unpause immediately.

* Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.h:
* Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.cpp:
(WebKit::ServiceWorkerDebuggableProxy::connect):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.h:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.messages.in:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::connectToInspector):
* Source/WebCore/workers/service/context/ServiceWorkerInspectorProxy.h:
* Source/WebCore/workers/service/context/ServiceWorkerInspectorProxy.cpp:
(WebCore::ServiceWorkerInspectorProxy::connectToWorker):
   - During step 3, when connecting the backend and frontend, pass in
     the values indicating auto inspection or pausing, and listen to the
     frontendInitialized event from the web process.

* Source/WebCore/workers/WorkerOrWorkletThread.h:
* Source/WebCore/workers/WorkerRunLoop.h:
   - Export these methods to be visible to WebKit.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/workers/service/context/ServiceWorkerDebuggable.h:
* Source/WebCore/workers/service/context/ServiceWorkerDebuggable.cpp:
(WebCore::ServiceWorkerDebuggable::connect):
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.h:
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
(WebCore::ServiceWorkerThreadProxy::ServiceWorkerThreadProxy):
(WebCore::ServiceWorkerThreadProxy::threadStartedRunningDebuggerTasks): Deleted.
   - As part of the migration, remove the auto-inspection support from
     the old ServiceWorkerDebuggable.

Canonical link: <a href="https://commits.webkit.org/291414@main">https://commits.webkit.org/291414@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/add9de851294a541450f972a2a3880515bd8e52d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92874 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/12425 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2071 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/97870 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/43400 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/94924 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12705 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20877 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/97870 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/43400 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95876 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/9575 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/84014 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/97870 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9268 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/1647 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/42713 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/85584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/79568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1620 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/99892 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/91540 "Built successfully and passed tests") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/19926 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14613 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/99892 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/20177 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79910 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/99892 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19694 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/1156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/19910 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/25086 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/114188 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/19597 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/32977 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/23057 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/21338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rcaliman-apple commented on Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
```diff
@@ -604,12 +628,20 @@ void WebSWContextManagerConnection::removeNavigationFetch(WebCore::SWServerConne
+            serviceWorkerThreadProxy->thread().runLoop().postDebuggerTask([serviceWorkerThreadProxy](ScriptExecutionContext&) {
+                serviceWorkerThreadProxy->thread().stopRunningDebuggerTasks();
+            });
+        };
```
👍 Looks a bit nicer now that you moved this logic here from `ServiceWorkerInspectorProxy::connectToWorker`. No more need to dig up a debuggable, which now lives in the UI process anyway.

rcaliman-apple commented on Source/JavaScriptCore/inspector/remote/RemoteInspectionTarget.cpp:
```diff
@@ -108,6 +108,8 @@ void RemoteInspectionTarget::pauseWaitingForAutomaticInspection()
             break;
     }
+
+    unpauseForInitializedInspector();
```
I guess this is replacing the removal from `RemoteInspector::updateAutomaticInspectionCandidate`

```diff
-    {
-       Locker locker { m_mutex };
-        m_pausedAutomaticInspectionCandidates.remove(targetIdentifier);
-    }
```

But is it still needed? 

As far as I can tell, this implementation is only used by `JSGlobalObjectDebuggable` (`ServiceWorkerDebuggableProxy` overrides `pauseWaitingForAutomaticInspection` and `unpauseForInitializedInspector `).

Unpause here does the same thing as `m_pausedAutomaticInspectionCandidates.remove(targetIdentifier);` which needs to be already done to break out of the `while` loop to get here.

That happens in either of the two cases:
- `JSGlobalObjectInspectorController::frontendInitialized()`
https://github.com/WebKit/WebKit/blob/326ff249dcc70b209836ede6e3cd95f30739e231/Source/JavaScriptCore/inspector/JSGlobalObjectInspectorController.cpp#L246-L248

- the 10 second timer expiring in `RemoteInspector::updateAutomaticInspectionCandidate`:
https://github.com/WebKit/WebKit/blob/326ff249dcc70b209836ede6e3cd95f30739e231/Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm#L203-L205

What use does additional `unpauseForInitializedInspector()` call serve?

rcaliman-apple replied:
Follow-up, if only `JSGlobalObjectDebuggable` uses these pause/unpause implementations defined in `RemoteInspectorTarget`, perhaps we should move them to the subclass.

the-chenergy replied:
Great catch! `unpause` needs to call the base class's version to remove the debuggable from the pending list. `unpause` doesn't do anything perceivable but just to keep variables consistent.

I think moving the busy-loop implementation to JSContext would be a good idea. In that case, we're letting any debuggable subclass explicitly define its behavior of pausing, rather than letting blocking be the default behavior. Although in that case, `pause` and `unpause` wouldn't be the best names anymore? But we don't have to worry about that now.

the-chenergy replied:
I simplified this part a bit to remove the need to override `unpause`. Let's think about refactoring the busy loop to be specialized by JSContext later.

the-chenergy replied:
Actually, the simplification I did was not completely correct. When in `RemoteInspector::receivedSetupMessage` and we're trying to set up the connection, we're relying on whether the debuggable object exists in that pending list to know whether auto-inspection was turned on! So, despite we don't want the debuggable to pause with a busy loop, we still want it to be kept in that pending list.

I'm working around that by [doing this for now](https://github.com/WebKit/WebKit/pull/41612/files#diff-0e8f6d1a6d234b9ad31c37877bcc875724a00d9e3d036dff8ca1b5c7545efc82) (the `FIXME`).

rcaliman-apple commented on Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
```diff
@@ -273,16 +273,17 @@
-    serviceWorkerDebuggableProxy->setInspectable(true);
-    serviceWorkerDebuggableProxy->init();
     m_serviceWorkerDebuggableProxies.add(identifier, serviceWorkerDebuggableProxy);
+    serviceWorkerDebuggableProxy->init();
+    serviceWorkerDebuggableProxy->setInspectable(true);
```
Thanks for taking care to order the calls correctly for automatic inspection 👍

rcaliman-apple commented on Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.cpp:
```diff
@@ -74,6 +74,18 @@ void ServiceWorkerDebuggableProxy::dispatchMessageFromRemote(String&& message)
+{
+    m_isPausedWaitingForAutomaticInspection = false;
+}
+#endif
```
Still need to call `RemoteInspector::singleton().setupCompleted()` to remove the debuggable from the `m_pausedAutomaticInspectionCandidates` and prevent the triggering of the 10 second timer set in `RemoteInspector::updateAutomaticInspectionCandidate`

pvollan commented on Source/WebCore/workers/service/context/ServiceWorkerDebuggable.cpp:
```diff
@@ -51,17 +51,10 @@ ServiceWorkerDebuggable::ServiceWorkerDebuggable(ServiceWorkerThreadProxy& servi
 }
 
-void ServiceWorkerDebuggable::connect(FrontendChannel& channel, bool isAutomaticInspection, bool immediatelyPause)
+void ServiceWorkerDebuggable::connect(FrontendChannel& channel, bool, bool)
```
Can these parameters be removed? I was also wondering if we can completely remove ServiceWorkerDebuggable? That can be done in a follow-up patch, if it is reasonable to remove it.

the-chenergy replied:
It turns out that I'd need to change some other less relevant call sites if those were removed. Let's do that later then, possibly by removing `ServiceWorkerDebuggable` altogether when the time comes.

pvollan reviewed (approved):
R=me.

patrickangle commented on Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.cpp:
```diff
@@ -74,6 +75,17 @@ void ServiceWorkerDebuggableProxy::dispatchMessageFromRemote(String&& message)
+{
+    m_wasRequestedToWaitForAutoInspection = true;
+
+    // This debuggable lives in the UI process, the same process receiving responses from the remote inspector,
```
This is only true sometimes. Only Safari has a UI Process that will be asked about inspection, but Service Workers can be originated from other processes as well (a Mac web app, for example)

the-chenergy committed:
Web Inspector: The new ServiceWorkerDebuggableProxy needs to adopt auto inspection
rdar://145795990
https://bugs.webkit.org/show_bug.cgi?id=288776

Reviewed by Patrick Angle and Per Arne Vollan.

The work in [1] introduced the ServiceWorkerDebuggableProxy that
provides debugging support to service worker from the UI process.
This commit migrates the work done in [2] to support automatically
inspecting this new debuggable type instead.

- [1] https://github.com/WebKit/WebKit/pull/40437 (Add ServiceWorkerDebuggableProxy)
- [2] https://github.com/WebKit/WebKit/pull/20316 (Enable auto-inspecting ServiceWorkerDebuggable)

To summarize how automatic inspection generally works:
1. The web content marks its debuggable as inspectable.
2. The debuggable blocks its thread before starting code execution.
3. The auto-launched frontend, after initializing, notifies the backend.
4. The backend unblocks the debuggable to resume its code execution.

The ServiceWorkerDebuggableProxy lives in the UI process but controls
the service worker's thread. Notably, during step 3, the UI process is
also the one handling responses from the remote inspector, meaning the
UI process can't be blocked while waiting for the frontend. Instead, it
tells the underlying service worker to remain paused. The web process
then carries on with the normal procedure of waiting for the frontend's
signal.

* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
(WebKit::WebProcessProxy::createServiceWorkerDebuggable):
* Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.cpp:
(WebKit::ServiceWorkerDebuggableProxy::pauseWaitingForAutomaticInspection):
   - In step 2, when we mark the debuggable as inspectable, it should
     pause and prevent code execution. This is now done as an async
     response to tell the service worker thread to remain paused from
     the WaitForInspector start mode.

* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::installServiceWorker):
   - To cooperate and finish step 2, handle the request to either pause
     or unpause immediately.

* Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.h:
* Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.cpp:
(WebKit::ServiceWorkerDebuggableProxy::connect):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.h:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.messages.in:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::connectToInspector):
* Source/WebCore/workers/service/context/ServiceWorkerInspectorProxy.h:
* Source/WebCore/workers/service/context/ServiceWorkerInspectorProxy.cpp:
(WebCore::ServiceWorkerInspectorProxy::connectToWorker):
   - During step 3, when connecting the backend and frontend, pass in
     the values indicating auto inspection or pausing, and listen to the
     frontendInitialized event from the web process.

* Source/WebCore/workers/WorkerOrWorkletThread.h:
* Source/WebCore/workers/WorkerRunLoop.h:
   - Export these methods to be visible to WebKit.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebCore/workers/service/context/ServiceWorkerDebuggable.h:
* Source/WebCore/workers/service/context/ServiceWorkerDebuggable.cpp:
(WebCore::ServiceWorkerDebuggable::connect):
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.h:
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
(WebCore::ServiceWorkerThreadProxy::ServiceWorkerThreadProxy):
(WebCore::ServiceWorkerThreadProxy::threadStartedRunningDebuggerTasks): Deleted.
   - As part of the migration, remove the auto-inspection support from
     the old ServiceWorkerDebuggable.

Canonical link: https://commits.webkit.org/291414@main
